### PR TITLE
k8s: update to 1.7.0

### LINKED ISSUE #975: Support Kubernetes v1.7.0
## Acceptance Criteria
- Users can install Kubernetes v1.7.0 on Linux and Windows clusters.

### PR DESCRIPTION
**Pull request by martell**
**What this PR does / why we need it**:
update to kubernetes 1.7.0

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #975

**Special notes for your reviewer**:
I tested this on Azure on eu-west.
Installed a bunch of stuff with helm and all was working as intended.
**Release note**:

```release-note
```
None

---
This change is [Reviewable](https://reviewable.io/reviews/azure/acs-engine/909)

### CODE REVIEW

File: pkg/acsengine/const.go [hunk 1]
```diff
@@ -54,6 +54,18 @@ const (
...
+		"hyperkube":    "hyperkube-amd64:v1.7.0",
+		"dashboard":    "kubernetes-dashboard-amd64:v1.6.1",
+		"exechealthz":  "exechealthz-amd64:1.2",
+		"addonresizer": "addon-resizer:2.0",
+		"heapster":     "heapster:v1.4.0",
+		"dns":          "k8s-dns-kube-dns-amd64:1.14.4",
+		"addonmanager": "kube-addon-manager-amd64:v6.1.2",
+		"dnsmasq":      "k8s-dns-dnsmasq-amd64:1.14.4",
+		"pause":        "pause-amd64:3.0",
+		"windowszip":   "v1.7.0intwinnat.zip",
```
Comment (JackQuincy) [inline pkg/acsengine/const.go]: @JiangtianLi I'm guessing this binary doesn't exist yet. Could you help with adding it?

File: pkg/api/const.go [hunk 2]
```diff
@@ -61,8 +61,10 @@ const (
 	Kubernetes162 OrchestratorVersion = "1.6.2"
 	// Kubernetes166 is the string constant for Kubernetes 1.6.6
 	Kubernetes166 OrchestratorVersion = "1.6.6"
+	// Kubernetes166 is the string constant for Kubernetes 1.6.6
+	Kubernetes170 OrchestratorVersion = "1.7.0"
 	// KubernetesLatest is the string constant for latest Kubernetes version
-	KubernetesLatest OrchestratorVersion = Kubernetes166
+	KubernetesLatest OrchestratorVersion = Kubernetes170
```
Comment (JackQuincy) [inline pkg/api/const.go]: while this is true it is the latest version we might want to mark 1.6.6 for a while as default until we have a little bake time on 1.7.0

Comment (martell) [inline pkg/api/const.go]: I was thinking that when I was doing the patch.
I believe this only occurs when you do not specify orchestratorVersion in your json.
Will update to reflect your comments.

File: pkg/acsengine/const.go (hunk 1 above)
Comment (JiangtianLi) [inline pkg/acsengine/const.go]: sure. we don't have 1.7.0 binary for Windows yet. we need to build and upload.

Comment (martell) [inline pkg/acsengine/const.go]: what can I say, I move fast waiting for no one :)

**Comment by JiangtianLi** [comment]
@martell v1.7.0intwinnat.zip has been uploaded. In kuberneteswindowssetup.ps1, please modify and remove `--enable-cri=false` for v1.7.0 and above

**Comment by JiangtianLi** [comment]
@martell please deploy a k8s windows 1.7.0 cluster and verify workload at https://github.com/Azure/acs-engine/blob/master/docs/kubernetes.windows.md works properly. Please let me know if you hit any issue. Thanks!

**Comment by JiangtianLi** [comment]
@martell Please merge properly with https://github.com/Azure/acs-engine/pull/892
@jackfrancis What do you suggest rate limit parameters for v1.7.0?

**Comment by jackfrancis** [comment]
@JiangtianLi `ratelimit*` and `backoff*` const values should be the same as `1.6.6`. I'll verify that k8s 1.7.0 hasn't changed any of the defaults for these const params:

- `nodestatusfreq`
- `nodegraceperiod`
- `podeviction`
- `routeperiod`

Stay tuned for final word on the above 4.

**Comment by jackfrancis** [comment]
@JiangtianLi Confirmed that none of the default config values for the params we're using have changed from k8s 1.6.6 to 1.7.0 (we're explicitly using those defaults as _our_ defaults in `const.go`). For clarity, here are the values and authoritative references in the k8s 1.7.0 release sha:

- nodestatusfreq: `"10s"`
  - https://github.com/kubernetes/kubernetes/blob/d3ada0119e776222f11ec7945e6d860061339aad/pkg/apis/componentconfig/v1alpha1/defaults.go#L313

- nodegraceperiod: `"40s"`
  - https://github.com/kubernetes/kubernetes/blob/d3ada0119e776222f11ec7945e6d860061339aad/cmd/kube-controller-manager/app/options/options.go#L87

- podeviction: `"5m0s"`
  - https://github.com/kubernetes/kubernetes/blob/d3ada0119e776222f11ec7945e6d860061339aad/cmd/kube-controller-manager/app/options/options.go#L86

- routeperiod: `"10s"`
  - https://github.com/kubernetes/kubernetes/blob/d3ada0119e776222f11ec7945e6d860061339aad/cmd/kube-controller-manager/app/options/options.go#L76

In summary, copy all values over from 1.6.6!

File: pkg/api/vlabs/const.go [hunk 3]
```diff
@@ -84,14 +84,16 @@ const (
...
 	// Kubernetes160 is the string constant for Kubernetes 1.6.0
 	Kubernetes160 OrchestratorVersion = "1.6.0"
 	// Kubernetes162 is the string constant for Kubernetes 1.6.2
 	Kubernetes162 OrchestratorVersion = "1.6.2"
 	// Kubernetes166 is the string constant for Kubernetes 1.6.6
 	Kubernetes166 OrchestratorVersion = "1.6.6"
+	// Kubernetes170 is the string constant for Kubernetes 1.7.0
+	Kubernetes170 OrchestratorVersion = "1.7.0"
 	// KubernetesLatest is the string constant for latest Kubernetes version
 	KubernetesLatest OrchestratorVersion = Kubernetes166
```
Comment (SaaldjorMike) [inline pkg/api/vlabs/const.go]: Wouldn't we also need to update this?

File: pkg/api/const.go [hunk 4]
```diff
@@ -61,6 +61,8 @@ const (
 	Kubernetes162 OrchestratorVersion = "1.6.2"
 	// Kubernetes166 is the string constant for Kubernetes 1.6.6
 	Kubernetes166 OrchestratorVersion = "1.6.6"
+	// Kubernetes166 is the string constant for Kubernetes 1.6.6
+	Kubernetes170 OrchestratorVersion = "1.7.0"
 	// KubernetesLatest is the string constant for latest Kubernetes version
 	KubernetesLatest OrchestratorVersion = Kubernetes166
```
Comment (SaaldjorMike) [inline pkg/api/const.go]: Wouldn't we also need to update this?

Comment (JackQuincy) [inline pkg/api/const.go]: We like to let a new version back a little while before making it default, possibly we should rename this.

Comment (SaaldjorMike) [inline pkg/api/const.go]: Makes sense. Was just wondering, due to the name of the variable + the variable description in the line above, that points to the fact that it is referring to the latest version and not the "default version" or whatever is a more suited name.

File: pkg/api/vlabs/const.go (hunk 3 above)
Comment (martell) [inline pkg/api/vlabs/const.go]: I previously did this but was asked not to so I did in a fixup.
This should be called `KubernetesDefault` instead.
With the go ahead from @JackQuincy I will rename that.

File: parts/kuberneteswindowssetup.ps1 [hunk 5]
```diff
@@ -175,7 +175,7 @@ Write-KubernetesStartFiles($podCIDR)
 c:\k\kubelet.exe --hostname-override=`$global:AzureHostname --pod-infra-container-image=kubletwin/pause --resolv-conf="" --allow-privileged=true --enable-debugging-handlers --api-servers=https://`${global:MasterIP}:443 --cluster-dns=`$global:KubeDnsServiceIp --cluster-domain=cluster.local  --kubeconfig=c:\k\config --hairpin-mode=promiscuous-bridge --v=2 --azure-container-registry-config=c:\k\azure.json
 "@
 
-    if ($global:KubeBinariesVersion -ne "1.5.3" -and $global:KubeBinariesVersion -ne "1.5.7")
+    if ($global:KubeBinariesVersion -ne "1.5.3" -and $global:KubeBinariesVersion -ne "1.5.7" -and $global:KubeBinariesVersion -lt "1.7.0")
```
Comment (JiangtianLi) [inline parts/kuberneteswindowssetup.ps1]: @martell Is `--image-pull-progress-deadline=20m --cgroups-per-qos=false --enforce-node-allocatable=`"`"` not used by 1.7.0? I think at least `--image-pull-progress-deadline=20m` is needed for downloading windows container image.

**Comment by JiangtianLi** [comment]
@martell Please merge master and test since there is a commit to fix failure in Windows node due to latest Windows version.

**Comment by JiangtianLi** [comment]
@martell The change LGTM. There is an issue with 1.7.0 zip for container internet connection due to that cri is now in 1.7.0 and can't be disabled. So I will need update zip and let you know.

**Comment by lachie83** [comment]
I also have a PR for 1.7.1 that builds on the awesome work that @martell has contributed. #1020 Which one shall we proceed with?

**Comment by martell** [comment]
@lachie83 seems like overkill to duplicate work :)
I rebased from master today to keep this updated.
I can bump the version to 1.7.1 if you really want to skip 1.7.0?

There are some extra thing in here
- rename KubernetesLatest to KubernetesCurrent
- fix typo in Kubernetes157 comment

@JiangtianLi any luck on your updated zips?
FYI I am using this in production quite happily.

**Comment by JiangtianLi** [comment]
@martell what kubelet/kubeproxy binaries are you using? The issue I have is that Windows container lack one NIC and doesn't have outbound connection. Do you see the same case?

**Comment by martell** [comment]
@JiangtianLi I am using this with linux containers currently.
I was waiting for updated zips before I tested windows containers.

**Comment by JiangtianLi** [comment]
@martell I see. Good to know Linux container is already working. Windows containers are built with additional changes. I'll try to solve the issue and update you.

File: examples/kubernetesversions/kubernetes1.7.0.json [hunk 6]
```diff
@@ -0,0 +1,36 @@
+{
+  "apiVersion": "vlabs",
+  "properties": {
+    "orchestratorProfile": {
+      "orchestratorType": "Kubernetes",
+      "orchestratorVersion": "1.7.0"
```
Comment (seanknox) [inline examples/kubernetesversions/kubernetes1.7.0.json]: I want to start removing cluster configs that provide little value—I'd opt to delete this one too.

File: examples/kubernetesversions/kubernetes1.7.1.json [hunk 7]
```diff
@@ -0,0 +1,36 @@
+{
+  "apiVersion": "vlabs",
+  "properties": {
+    "orchestratorProfile": {
+      "orchestratorType": "Kubernetes",
+      "orchestratorVersion": "1.7.1"
```
Comment (seanknox) [inline examples/kubernetesversions/kubernetes1.7.1.json]: Remove this example too please.

File: parts/kuberneteswindowssetup.ps1 [hunk 8]
```diff
@@ -178,10 +178,14 @@ Write-KubernetesStartFiles($podCIDR)
 c:\k\kubelet.exe --hostname-override=`$global:AzureHostname --pod-infra-container-image=kubletwin/pause --resolv-conf="" --allow-privileged=true --enable-debugging-handlers --api-servers=https://`${global:MasterIP}:443 --cluster-dns=`$global:KubeDnsServiceIp --cluster-domain=cluster.local  --kubeconfig=c:\k\config --hairpin-mode=promiscuous-bridge --v=2 --azure-container-registry-config=c:\k\azure.json --runtime-request-timeout=10m
 "@
 
-    if ($global:KubeBinariesVersion -ne "1.5.3" -and $global:KubeBinariesVersion -ne "1.5.7")
+    if ($global:KubeBinariesVersion -ne "1.5.3" -and $global:KubeBinariesVersion -ne "1.5.7")
     {
-        $KubeletArgList += "--enable-cri=false"
-        $KubeletCommandLine += " --enable-cri=false --image-pull-progress-deadline=20m --cgroups-per-qos=false --enforce-node-allocatable=`"`""
+        if ($global:KubeBinariesVersion -lt "1.7.0")
```
Comment (seanknox) [inline parts/kuberneteswindowssetup.ps1]: This is unclear to me; what versions are we trying to target for this conditional?  Can this be rewritten to be something like

```powershell
 if ($global:KubeBinariesVersion -ge "1.6.0" -and $global:KubeBinariesVersion -le "1.7.0")
```

File: pkg/acsengine/engine.go [hunk 9]
```diff
@@ -850,7 +850,7 @@ func (t *TemplateGenerator) getTemplateFuncMap(cs *api.ContainerService) map[str
 				case "kubeClusterCidr":
 					val = "10.244.0.0/16"
 				case "kubeBinariesVersion":
-					val = string(api.KubernetesLatest)
+					val = string(api.KubernetesCurrent)
```
Comment (seanknox) [inline pkg/acsengine/engine.go]: If you're going to change this, I'd opt for something clearer like `KubernetesDefaultVersion`, or just not change it in this PR.

File: pkg/acsengine/const.go [hunk 10]
```diff
@@ -80,6 +80,50 @@ const (
 
 // KubeImages represents Docker images used for Kubernetes components based on Kubernetes version
 var KubeImages = map[api.OrchestratorVersion]map[string]string{
+	api.Kubernetes171: {
+		"hyperkube":    "hyperkube-amd64:v1.7.1",
+		"dashboard":    "kubernetes-dashboard-amd64:v1.6.1",
```
Comment (seanknox) [inline pkg/acsengine/const.go]: This PR started out as v1.7.0, and I'd like it to continue to be focused on that. Each Kubernetes version needs to be vetted, run through our CI, Windows artifacts created, etc.

File: examples/kubernetesversions/kubernetes1.7.0.json (hunk 6 above)
Comment (martell) [inline examples/kubernetesversions/kubernetes1.7.0.json]: Sure can do

File: examples/kubernetesversions/kubernetes1.7.1.json (hunk 7 above)
Comment (martell) [inline examples/kubernetesversions/kubernetes1.7.1.json]: sure

File: pkg/acsengine/engine.go (hunk 9 above)
Comment (martell) [inline pkg/acsengine/engine.go]: done

File: pkg/acsengine/const.go (hunk 10 above)
Comment (martell) [inline pkg/acsengine/const.go]: Was not aware of that process.
Will do

File: parts/kuberneteswindowssetup.ps1 (hunk 8 above)
Comment (martell) [inline parts/kuberneteswindowssetup.ps1]: sure

**Comment by martell** [comment]
@seanknox I wasn't aware of the internal testing procedures involved with vetting each version.
More then happy to have a single PR for each. My concern with having 2 PR's was that 1.7.0 would be skipped when I was running into intermittent issues like `installation: Error: client: etcd cluster is unavailable or misconfigured` on 1.7.1.
I assume the windows issues @JiangtianLi has with the NAT is external to acs-engine source code.
With the default still being 1.6.6 I assume it should be possible to merge this?
The quicker this goes in the quicker @lachie83 can rebase the new PR off master and to get 1.7.1  internal testing kickstarted. :)

File: parts/kuberneteswindowssetup.ps1 [hunk 11]
```diff
@@ -178,10 +178,14 @@ Write-KubernetesStartFiles($podCIDR)
...
+    if ($global:KubeBinariesVersion -ge "1.6.0" -and $global:KubeBinariesVersion -le "1.7.0")
+    {
+            $KubeletArgList += "--enable-cri=false"
+            $KubeletCommandLine += " --enable-cri=false"
+    }
     if ($global:KubeBinariesVersion -ne "1.5.3" -and $global:KubeBinariesVersion -ne "1.5.7")
     {
-        $KubeletArgList += "--enable-cri=false"
-        $KubeletCommandLine += " --enable-cri=false --image-pull-progress-deadline=20m --cgroups-per-qos=false --enforce-node-allocatable=`"`""
+        $KubeletCommandLine += " --image-pull-progress-deadline=20m --cgroups-per-qos=false --enforce-node-allocatable=`"`""
```
Comment (seanknox) [inline parts/kuberneteswindowssetup.ps1]: @martell or @JiangtianLi can you you add a comment above these conditionals so others understand what they do (and why we have them)?

**Comment by seanknox** [comment]
@acs-bot test this please

**Comment by seanknox** [comment]
@martell No worries at all and thanks for your contribution! This looks LGTM, have one request to be document the need for different kubelet flags in `parts/kuberneteswindowssetup.ps1`.

File: parts/kuberneteswindowssetup.ps1 (hunk 11 above)
Comment (martell) [inline parts/kuberneteswindowssetup.ps1]: For clarification I only removed `--enable-cri=false` from these 2 lines to guard it from 1.6x+
We can probably simplify the conditional to just `if ($global:KubeBinariesVersion -ge "1.6.0")`?
and the comment would simply be `we don't use Container Runtime Interface from 1.6.0+`
Would that suffice @seanknox ?

**Comment by JiangtianLi** [comment]
@seanknox @martell --enable-cri is deprecated from 1.7.0. Other options has good explanation in cmd/kubelet/app/options/options.go as we need them for the Window path, i.e., without them, Windows code path are not implemented. --image-pull-progress-deadline=20m is due to full Windows Server image size.

File: parts/kuberneteswindowssetup.ps1 (hunk 11 above)
Comment (seanknox) [inline parts/kuberneteswindowssetup.ps1]: FYI, I want to provide as much context to users here without forcing them to dig through Kubernetes source code.

Based on @JiangtianLi comment below, `--enable-cri` is deprecated starting in v1.7.0. I think just a comment to that effect is fine.

`--image-pull-progress-deadline=20m` is needed due to the size of the container images.  This option is enabled only for k8s versions ~v1.5.3 to ~v1.5.7— @JiangtianLi is that intentional or should this be on for all versions?

Comment (JiangtianLi) [inline parts/kuberneteswindowssetup.ps1]: @seanknox yes, 1.5.* doesn't have --image-pull-progress-deadline option. The timeout is hardcoded. We have that option from 1.6.0 and above.

Comment (martell) [inline parts/kuberneteswindowssetup.ps1]: done

File: parts/kuberneteswindowssetup.ps1 [hunk 12]
```diff
@@ -177,11 +177,16 @@ Write-KubernetesStartFiles($podCIDR)
     $KubeletCommandLine = @"
 c:\k\kubelet.exe --hostname-override=`$global:AzureHostname --pod-infra-container-image=kubletwin/pause --resolv-conf="" --allow-privileged=true --enable-debugging-handlers --api-servers=https://`${global:MasterIP}:443 --cluster-dns=`$global:KubeDnsServiceIp --cluster-domain=cluster.local  --kubeconfig=c:\k\config --hairpin-mode=promiscuous-bridge --v=2 --azure-container-registry-config=c:\k\azure.json --runtime-request-timeout=10m
 "@
-
+    # stop using container runtime interface from 1.6.0+ (officially deprecated from 1.7.0)
+    if ($global:KubeBinariesVersion -ge "1.6.0")
```
Comment (seanknox) [inline parts/kuberneteswindowssetup.ps1]: Think you want this to be:
```powershell
if ($global:KubeBinariesVersion -ge "1.6.0" -and $global:KubeBinariesVersion -lt "1.7.0")
```

That ensures users are using between v1.6.0 up until v1.7.0

File: parts/kuberneteswindowssetup.ps1 (hunk 11 above)
Comment (seanknox) [inline parts/kuberneteswindowssetup.ps1]: Ok, so `image-pull-progress-deadline` requires v1.6.0 at a minimum. @martell you should update that logic to be

```powershell
if ($global:KubeBinariesVersion -ge "1.6.0")
```

No need to bookend k8s versions here as it continues to be supported in v1.7.0.

File: parts/kuberneteswindowssetup.ps1 (hunk 12 above)
Comment (martell) [inline parts/kuberneteswindowssetup.ps1]: The flag would still be valid for 1.7.0. The deprecation just makes sure this happens by default. I suppose it might become an invalid flag in 1.8+

Comment (seanknox) [inline parts/kuberneteswindowssetup.ps1]: Fair enough. In that case, I think you could group both checks under one check for v1.6.0 or greater.

**Comment by martell** [comment]
@seanknox done :)

**Comment by seanknox** [comment]
@acs-bot test this please

**Comment by andrey-moor** [comment]
Would it be possible to upgrade from 1.6 without deploying a new cluster?

**Comment by amanohar** [comment]
@andrey-moor ACS Engine today does have an experimental upgrade feature in the work that will enable upgrade with redeploy (but will recreate agent and master nodes). At some point ACS Engine will support upgrade to 1.7.0.

**Comment by seanknox** [comment]
Thanks again @martell @JiangtianLi!

**Comment by seanknox** [comment]
giphy-downsized 4
